File: src/python/bashkit_ipc_driver.py
"""BashkitIPCDriver: JSON-RPC over stdin/stdout with bashkit-cli."""

from __future__ import annotations

import json
import subprocess
from typing import Any, Callable

from src.python.drivers import ShellDriver, FilesystemDriver, BuiltinFilesystemDriver
from src.python.shell import ExecResult


class BashkitIPCDriver(ShellDriver):
    """ShellDriver that communicates with bashkit-cli via JSON-RPC over stdio."""

    def __init__(self, cwd: str = "/", env: dict[str, str] | None = None, **kwargs: Any):
        self._cwd = cwd
        self._env = env if env is not None else {}
        self._fs_driver = BuiltinFilesystemDriver()
        self._commands: dict[str, Callable] = {}
        self._on_not_found: Callable | None = None
        self._request_id = 0
        self._process = self._spawn()

    def _spawn(self) -> Any:
        """Spawn the bashkit-cli process. Overridden in tests."""
        return subprocess.Popen(
            ["bashkit-cli", "--jsonrpc"],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )

    def _next_id(self) -> int:
        self._request_id += 1
        return self._request_id

    def _send(self, msg: dict) -> None:
        """Write a JSON-RPC message to the process stdin."""
        self._process.stdin.write(json.dumps(msg) + "\n")
        self._process.stdin.flush()

    def _recv(self) -> dict:
        """Read a JSON-RPC message from the process stdout."""
        line = self._process.stdout.readline()
        if not line:
            return {}
        return json.loads(line.strip())

    def _snapshot_fs(self) -> dict[str, str]:
        """Serialize all VFS files to a dict, resolving lazy providers."""
        snapshot: dict[str, str] = {}
        for path in self._fs_driver.find("/", "*"):
            if not self._fs_driver.is_dir(path):
                snapshot[path] = self._fs_driver.read_text(path)
        return snapshot

    def _apply_fs_changes(self, changes: dict) -> None:
        """Apply created files and deleted paths back to the host FS."""
        created = changes.get("created", {})
        deleted = changes.get("deleted", [])
        for path, content in created.items():
            self._fs_driver.write(path, content)
        for path in deleted:
            if self._fs_driver.exists(path):
                self._fs_driver.remove(path)

    @property
    def fs(self) -> FilesystemDriver:
        return self._fs_driver

    @property
    def cwd(self) -> str:
        return self._cwd

    @property
    def env(self) -> dict[str, str]:
        return self._env

    @property
    def on_not_found(self) -> Callable | None:
        return self._on_not_found

    @on_not_found.setter
    def on_not_found(self, callback: Callable | None) -> None:
        self._on_not_found = callback

    def exec(self, command: str) -> ExecResult:
        """Execute a command via bashkit-cli JSON-RPC."""
        req_id = self._next_id()
        snapshot = self._snapshot_fs()

        self._send({
            "id": req_id,
            "method": "exec",
            "params": {
                "cmd": command,
                "cwd": self._cwd,
                "env": self._env,
                "fs": snapshot,
            },
        })

        # Event loop: handle callbacks until we get the final result
        while True:
            response = self._recv()
            if not response:
                return ExecResult(stderr="No response from bashkit-cli", exit_code=1)

            # Error response
            if "error" in response:
                error = response["error"]
                msg = error.get("message", "Unknown error") if isinstance(error, dict) else str(error)
                return ExecResult(stderr=msg, exit_code=1)

            # Callback: invoke_command from bashkit
            if response.get("method") == "invoke_command":
                cb_id = response.get("id")
                params = response.get("params", {})
                name = params.get("name", "")
                args = params.get("args", [])
                stdin = params.get("stdin", "")

                if name in self._commands:
                    try:
                        result = self._commands[name](args, stdin=stdin)
                        self._send({"id": cb_id, "result": result})
                    except Exception as e:
                        self._send({"id": cb_id, "error": {"code": -1, "message": str(e)}})
                else:
                    self._send({"id": cb_id, "error": {"code": -1, "message": f"Unknown command: {name}"}})
                continue

            # Final result
            if "result" in response:
                r = response["result"]
                if "fs_changes" in r:
                    self._apply_fs_changes(r["fs_changes"])
                return ExecResult(
                    stdout=r.get("stdout", ""),
                    stderr=r.get("stderr", ""),
                    exit_code=r.get("exitCode", 0),
                )

    def register_command(self, name: str, handler: Callable) -> None:
        """Register a custom command handler locally and notify bashkit-cli."""
        self._commands[name] = handler
        self._send({
            "method": "register_command",
            "params": {"name": name},
        })

    def unregister_command(self, name: str) -> None:
        """Remove a custom command handler and notify bashkit-cli."""
        self._commands.pop(name, None)
        self._send({
            "method": "unregister_command",
            "params": {"name": name},
        })

    def clone(self) -> BashkitIPCDriver:
        """Create a new independent instance with cloned FS and same config."""
        new_driver = BashkitIPCDriver(cwd=self._cwd, env=dict(self._env))
        # Clone the filesystem
        cloned_fs = self._fs_driver.clone()
        new_driver._fs_driver = cloned_fs
        # Copy registered commands and re-register with the new process
        new_driver._commands = dict(self._commands)
        new_driver._on_not_found = self._on_not_found
        for name in self._commands:
            new_driver._send({
                "method": "register_command",
                "params": {"name": name},
            })
        return new_driver

    def __del__(self) -> None:
        """Terminate the bashkit-cli process on cleanup."""
        proc = getattr(self, "_process", None)
        if proc is not None:
            try:
                proc.terminate()
                proc.wait(timeout=5)
            except Exception:
                pass

File: src/typescript/bashkit-ipc-driver.ts
/**
 * BashkitIPCDriver: JSON-RPC over stdin/stdout with bashkit-cli.
 */

import {
  type ShellDriver,
  type FilesystemDriver,
  type ShellDriverOptions,
  BuiltinFilesystemDriver,
} from "./drivers.js";
import type { ExecResult, CmdHandler } from "./shell.js";

export interface BashkitIPCDriverOptions extends ShellDriverOptions {
  _spawnOverride?: () => any;
}

interface JsonRpcMessage {
  id?: number;
  method?: string;
  params?: Record<string, unknown>;
  result?: Record<string, unknown>;
  error?: { code: number; message: string } | string;
}

export class BashkitIPCDriver implements ShellDriver {
  private _cwd: string;
  private _env: Record<string, string>;
  private _fsDriver: BuiltinFilesystemDriver;
  private _commands: Map<string, CmdHandler> = new Map();
  private _onNotFound?: (cmdName: string) => void;
  private _requestId = 0;
  private _process: any;

  constructor(opts: BashkitIPCDriverOptions = {}) {
    this._cwd = opts.cwd ?? "/";
    this._env = opts.env ?? {};
    this._fsDriver = new BuiltinFilesystemDriver();
    this._process = opts._spawnOverride
      ? opts._spawnOverride()
      : this._spawn();
  }

  private _spawn(): any {
    // For real usage, spawn a persistent bashkit-cli process.
    // Currently a placeholder — real IPC transport is a future concern.
    throw new Error(
      "bashkit-cli not available. Use _spawnOverride for testing."
    );
  }

  get fs(): FilesystemDriver {
    return this._fsDriver;
  }

  get cwd(): string {
    return this._cwd;
  }

  get env(): Record<string, string> {
    return this._env;
  }

  get onNotFound(): ((cmdName: string) => void) | undefined {
    return this._onNotFound;
  }

  set onNotFound(cb: ((cmdName: string) => void) | undefined) {
    this._onNotFound = cb;
  }

  private _nextId(): number {
    this._requestId += 1;
    return this._requestId;
  }

  private _send(msg: Record<string, unknown>): void {
    this._process.stdin.write(JSON.stringify(msg) + "\n");
  }

  private _recv(): JsonRpcMessage {
    const line: string = this._process.stdout.readline();
    if (!line) return {};
    return JSON.parse(line.trim());
  }

  private _snapshotFs(): Record<string, string> {
    const snapshot: Record<string, string> = {};
    for (const path of this._fsDriver.find("/", "*")) {
      if (!this._fsDriver.isDir(path)) {
        snapshot[path] = this._fsDriver.readText(path);
      }
    }
    return snapshot;
  }

  private _applyFsChanges(changes: Record<string, unknown>): void {
    const created = (changes.created ?? {}) as Record<string, string>;
    const deleted = (changes.deleted ?? []) as string[];
    for (const [path, content] of Object.entries(created)) {
      this._fsDriver.write(path, content);
    }
    for (const path of deleted) {
      if (this._fsDriver.exists(path)) {
        this._fsDriver.remove(path);
      }
    }
  }

  exec(command: string): ExecResult {
    const reqId = this._nextId();
    const snapshot = this._snapshotFs();

    this._send({
      id: reqId,
      method: "exec",
      params: {
        cmd: command,
        cwd: this._cwd,
        env: this._env,
        fs: snapshot,
      },
    });

    // Event loop: handle callbacks until we get the final result
    while (true) {
      const response = this._recv();
      if (!response || Object.keys(response).length === 0) {
        return { stdout: "", stderr: "No response from bashkit-cli", exitCode: 1 };
      }

      // Error response
      if (response.error !== undefined) {
        const error = response.error;
        const msg =
          typeof error === "object" && error !== null
            ? (error as any).message ?? "Unknown error"
            : String(error);
        return { stdout: "", stderr: msg, exitCode: 1 };
      }

      // Callback: invoke_command from bashkit
      if (response.method === "invoke_command") {
        const cbId = response.id;
        const params = response.params ?? {};
        const name = (params.name as string) ?? "";
        const args = (params.args as string[]) ?? [];
        const stdin = (params.stdin as string) ?? "";

        if (this._commands.has(name)) {
          try {
            const result = this._commands.get(name)!(args, stdin);
            this._send({ id: cbId, result });
          } catch (e: unknown) {
            const errMsg = e instanceof Error ? e.message : String(e);
            this._send({
              id: cbId,
              error: { code: -1, message: errMsg },
            });
          }
        } else {
          this._send({
            id: cbId,
            error: { code: -1, message: `Unknown command: ${name}` },
          });
        }
        continue;
      }

      // Final result
      if (response.result !== undefined) {
        const r = response.result as Record<string, unknown>;
        if (r.fs_changes) {
          this._applyFsChanges(r.fs_changes as Record<string, unknown>);
        }
        return {
          stdout: (r.stdout as string) ?? "",
          stderr: (r.stderr as string) ?? "",
          exitCode: (r.exitCode as number) ?? 0,
        };
      }
    }
  }

  registerCommand(name: string, handler: CmdHandler): void {
    this._commands.set(name, handler);
    this._send({
      method: "register_command",
      params: { name },
    });
  }

  unregisterCommand(name: string): void {
    this._commands.delete(name);
    this._send({
      method: "unregister_command",
      params: { name },
    });
  }

  clone(opts?: { _spawnOverride?: () => any }): BashkitIPCDriver {
    const spawnOverride = opts?._spawnOverride;
    const newDriver = new BashkitIPCDriver({
      cwd: this._cwd,
      env: { ...this._env },
      _spawnOverride: spawnOverride,
    });

    // Clone the filesystem
    newDriver._fsDriver = this._fsDriver.clone();

    // Copy registered commands and re-register with new process
    newDriver._commands = new Map(this._commands);
    newDriver._onNotFound = this._onNotFound;

    for (const name of this._commands.keys()) {
      newDriver._send({
        method: "register_command",
        params: { name },
      });
    }

    return newDriver;
  }

  destroy(): void {
    if (this._process) {
      try {
        this._process.kill();
      } catch {
        // Ignore errors during cleanup
      }
    }
  }
}
